cpp/inc/bond/ext/grpc/detail/service_call_data.h

```diff
@@ -81,30 +84,26 @@ struct service_unary_call_data : io_manager_tag
     {
         if (ok)
         {
-            // capture the data associated with this one incomming request
-            // so that we can pass it to the user callback. The unary_call
-            // that we create to pass to the user callback takes ownership
-            // of this data.
-            //
-            // This scope block is here so that the shared_ptr that we
-            // create to capture in the lambda is destroyed after we enqueue
-            // in the thread pool but before we schedule the next receive.
-            // When we can use C++14, we can simplify this by using lambda's
-            // capture-by-move.
+            // Capture the data associated with this one incomming request
+            // so that we can pass it to the user callback. When we create
+            // the unary_call to pass to the user callback, the unary_call
+            // and unary_call_impl start collaborating to manage the
+            // lifetime of the unary_call_impl.
             {
-                std::shared_ptr<unary_call_impl<TRequest, TResponse>> receivedCall = _receivedCall;
+                // We have to manually release the pointer from
+                // _receivedCall to capture it in the lambda. When we can
+                // use C++14, we can simplify this by using lambda's
+                // capture-by-move.
+                auto receivedCall = boost::intrusive_ptr<uc_impl>{ _receivedCall.release() };
```

Comment:
Should `_receivedCall` be changed to `boost::intrusive_ptr` as well?

Reply:
Not easily that I can see. `uc_impl::Release` is responsible for the guaranteed send behavior, but that is only needed _after_ a call has been received. By keeping the `uc_impl` in a `unique_ptr` in `service_unary_call_data`, if a call is never received, the we just delete the `uc_impl` when the service instance is destroyed. Its only when a call has been received that we start refcounting and triggering the guaranteed send behavior.

I tried moving this has-call-been-received behavior inside of `uc_impl`, but it got gross quickly.